### PR TITLE
handle semver git urls. fix https://github.com/pnpm/pnpm/issues/1071

### PR DESCRIPTION
**Pull request by andreineculau**
ping @zkochan 

---

sidenote on the dev process: this is a really small repo, but what a slow experience it was

1. running the tests (`npm run test`) shaves off around 20s on my fast mac just to run 1 basic test
2. it is impossible to run the tests alone
3. my shrinkwrap.yml is dirty after running `npm run test` due to `preview`

### CODE REVIEW

File: src/parsePref.ts [hunk 1]
```diff
@@ -12,12 +12,9 @@ export type HostedPackageSpec = ({
...
   },
   normalizedPref: string,
-} & ({
   gitCommittish: string | null,
-} | {
-  gitCommittish: null,
-  gitRange: string,
-}))
+  gitRange?: string,
+})
```
Comment (andreineculau) [inline src/parsePref.ts]: i'm no typescript lover, and without these changes, I was upsetting typescript
i can change it to whatever pleases typescript and the maintainers

```
src/index.ts(32,110): error TS2339: Property 'gitRange' does not exist on type '({ fetchSpec: string; hosted?: { type: string; shortcut: string; sshUrl: string; user: string; pr...'.
  Property 'gitRange' does not exist on type '{ fetchSpec: string; hosted?: { type: string; shortcut: string; sshUrl: string; user: string; pro...'.
src/index.ts(56,37): error TS2339: Property 'gitRange' does not exist on type '({ fetchSpec: string; hosted?: { type: string; shortcut: string; sshUrl: string; user: string; pr...'.
  Property 'gitRange' does not exist on type '{ fetchSpec: string; hosted?: { type: string; shortcut: string; sshUrl: string; user: string; pro...'.
src/index.ts(68,92): error TS2339: Property 'gitRange' does not exist on type '({ fetchSpec: string; hosted?: { type: string; shortcut: string; sshUrl: string; user: string; pr...'.
  Property 'gitRange' does not exist on type '{ fetchSpec: string; hosted?: { type: string; shortcut: string; sshUrl: string; user: string; pro...'.
src/index.ts(71,90): error TS2339: Property 'gitRange' does not exist on type '({ fetchSpec: string; hosted?: { type: string; shortcut: string; sshUrl: string; user: string; pr...'.
  Property 'gitRange' does not exist on type '{ fetchSpec: string; hosted?: { type: string; shortcut: string; sshUrl: string; user: string; pro...'.
```

**Comment by andreineculau** [comment]
FWIW I have no idea why Node.js v4 fails https://travis-ci.org/pnpm/git-resolver/jobs/352310760
PS: it has end of life 2018-04-30

File: src/index.ts [hunk 2]
```diff
@@ -81,15 +82,48 @@ export default function (
...
+function resolveVTags (vTags: string[], range: string) {
+  return semver.maxSatisfying(vTags, range, true)
+}
 
-  // if no ref was found, assume that ref is the commit ID
-  if (!result.stdout) return ref
+async function resolveRef (repo: string, ref: string, range?: string) {
+  const result = await git(['ls-remote', '--refs', repo])
+  const refs = result.stdout.split('\n').reduce((obj: object, line: string) => {
+    const [commit, commitRef] = line.split('\t')
```
Comment (zkochan) [inline src/index.ts]: this is why it fails on Node 4. destructuring

**Comment by zkochan** [comment]
Thanks for your contribution! I will review it tonight

File: src/parsePref.ts (hunk 1 above)
Comment (zkochan) [reply src/parsePref.ts]: ok, maybe it's fine, I'll check after merging

**Comment by andreineculau** [comment]
PR updated with all the issues you pointed out resolved. Thanks

**Pull request merged**